tracing/src/main/java/com/palantir/tracing/DetachedSpan.java

```diff
@@ -48,6 +48,15 @@ static DetachedSpan start(String operation, SpanType type) {
         return Tracer.detachInternal(operation, type);
     }
 
+    /**
+     * Converts the current in-progress thread-local span into a {@link DetachedSpan}, which can be used to more easily
+     * instrument async operations. This {@link DetachedSpan} will need to be manually closed.
+     */
+    @CheckReturnValue
+    static DetachedSpan detachThreadLocalSpan() {
+        return Tracer.detachThreadLocalSpan();
+    }
```

Comment:
I think this is a foot-gun. It's easy to refactor the start/complete attached span pieces without remembering to remove an detachThreadLocalSpan or migrate it to  create a new detached span.

Reply:
@carterkozak would love to know what you suggest here. #316 is the issue we were trying to solve.

Reply:
A couple potential options (neither of which is ideal):
1. We create child spans for subsequent operations. The output may not be pretty, but it will give us more information about the processing flow.
2. Create a single detached span for the duration of the call. It's not attached to a thread, so logging will not be associated with the traceId, but we get pretty traces without risking clobbering thread state.

Is the delta between existing options and the ideal worth the risk of corrupted tracing?

Reply:
Let me explain a little bit more what I'd like to do so we are on the same page.

I have a function that does something like

```
ListenableFuture<A> valueA = callA();
ListenableFuture<B> valueB = callB();
ListenableFuture<C> valueC = callC();
```

All of those functions `callA`, `callB`, `callC` do some work synchronously and produce traces. I would like to have all calls to those functions have distinct so imagine I would wrap all those functions calls in something like

```
Tracer.fastStartSpan("call to A");
ListenableFuture<A> valueA = callA();
DetachedSpan span = Tracer.detachThreadLocalSpan();
Futures.addCallback(result, new FutureCallback<>() {
    @Override
    public void onSuccess(@Nullable TemporaryCredentialsAuthToken result) {
        span.complete();
    }

    @Override
    public void onFailure(Throwable t) {
        span.complete();
    }
}, MoreExecutors.directExecutor());
return result;
```
After thinking about how to deal with this I couldn't figure out how to do it without detach span. 

Creating a detached span for duration of the call is confusing since you don't know parent relations and you only use the start times as a proxy which leads to confusing traces

Reply:
Does this work?

```java
DetachedSpan detached = DetachedSpan.start("call to A");
ListenableFuture<A> valueA;
try (CloseableSpan span = detached.childSpan("Initiate call to A")) {
    valueA = callA();
}
valueA.addListener(detached::complete, MoreExecutors.directExecutor());
```

Reply:
I thought there was something weird about this one but let me try again